### PR TITLE
Add function to save point cloud in xdmf

### PR DESCRIPTION
**Pull request by finsberg**
Address https://github.com/scientificcomputing/scifem/issues/9

### CODE REVIEW

File: src/scifem/xdmf.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    adios2 = resolve_adios_scope(adios2)
+
+    u = us[0]
+    points = u.function_space.tabulate_dof_coordinates()
+    h5name = Path(filename).with_suffix(".h5")
+
+    bs = u.function_space.dofmap.index_map_bs
+    comm = u.function_space.mesh.comm
+    num_dofs_global = u.function_space.dofmap.index_map.size_global
+    num_dofs_local = u.function_space.dofmap.index_map.size_local
```
Comment (jorgensd) [inline src/scifem/xdmf.py:40]:
```suggestion
    dofmap =  V.dofmap
    imap = dofmap.index_map
    bs = dofmap.index_map_bs
    comm = V.mesh.comm
    num_dofs_global = imap.size_global
    num_dofs_local = imap.size_local
```

File: src/scifem/xdmf.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    import adios2
+
+    def resolve_adios_scope(adios2):
+        return adios2 if not hasattr(adios2, "bindings") else adios2.bindings
+
+    adios2 = resolve_adios_scope(adios2)
+
+    u = us[0]
+    points = u.function_space.tabulate_dof_coordinates()
```
Comment (jorgensd) [inline src/scifem/xdmf.py:34]:
```suggestion
    u = us[0]
    V = u.function_space
    points = V.tabulate_dof_coordinates()
```

File: src/scifem/xdmf.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    h5name = Path(filename).with_suffix(".h5")
+
+    bs = u.function_space.dofmap.index_map_bs
+    comm = u.function_space.mesh.comm
+    num_dofs_global = u.function_space.dofmap.index_map.size_global
+    num_dofs_local = u.function_space.dofmap.index_map.size_local
+    local_range = np.array(u.function_space.dofmap.index_map.local_range, dtype=np.int64)
```
Comment (jorgensd) [inline src/scifem/xdmf.py:41]:
```suggestion
    local_range = np.array(imap..local_range, dtype=np.int64)
```

File: src/scifem/xdmf.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            it1.text = f"{h5name.name}:/Step0/Values_{u.name}"
+        text = [
+            '<?xml version="1.0"?>\n<!DOCTYPE Xdmf SYSTEM "Xdmf.dtd" []>\n',
+            ET.tostring(xdmf, encoding="unicode"),
+        ]
+        Path(filename).with_suffix(".xdmf").write_text("".join(text))
+    # Create ADIOS2 reader
+    adios = adios2.ADIOS(comm)
+    io = adios.DeclareIO("Point cloud writer")
+    io.SetEngine("HDF5")
```
Comment (jorgensd) [inline src/scifem/xdmf.py:82]: Should we here do something similar to:
https://github.com/jorgensd/mesh_converter/blob/main/src/mesh_converter/xdmf.py#L254-L294
where we support adios2 from `pip` (built without MPI), as well as having a fallback to `h5py` if adios2 Python interface is not available?

File: src/scifem/xdmf.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    pointvar = io.DefineVariable(
+        "Points",
+        points_out,
+        shape=[num_dofs_global, points.shape[1]],
+        start=[local_range[0], 0],
+        count=[num_dofs_local, points.shape[1]],
+    )
+    outfile.Put(pointvar, points_out)
+    for u in us:
```
Comment (jorgensd) [inline src/scifem/xdmf.py:94]: Should we have some sanity checks on `us?` That all function spaces are the same?

**Review by jorgensd — CHANGES_REQUESTED** [review]
See various comments. We also need to add more dependencies to pyproject (optional, `[h5]` or `[adios2]` for write packages.

File: src/scifem/xdmf.py (hunk 4 above)
Comment (finsberg) [reply src/scifem/xdmf.py:85]: Good idea! I will make a suggestion

**Pull request merged**